app-libs/stf/src/challenge_code.rs

```diff
@@ -0,0 +1,95 @@
+/*
+	Copyright 2022 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use rand::Rng;
+
+pub trait ChallengeCodeGenerator {
+	fn generate(&mut self) -> String;
+}
+
+#[derive(Default)]
+pub struct NumericChallengeCode {
```

Comment:
We could use something like [this](https://rust-lang-nursery.github.io/rust-cookbook/algorithms/randomness.html#create-random-passwords-from-a-set-of-alphanumeric-characters) to make it more concise and less error-prone

We could start with the default one (e.g. mixed of alhpa and numeric chars) and change it if we have other needs.

We could also simplify the struct as something like
```
pub struct AlphaNumbericChallengeCodeGenerator(pub len: u8);
```
The name is a bit too long..

Reply:
Missed the second half of the cookbook examples.